src/System.Threading/tests/AsyncLocalTests.cs

```diff
@@ -419,5 +423,162 @@ private static Task Run(Func<Task> func)
+                    {
+                        Assert.True(false, "CurrentValue and PreviousValue are both set");
+                    }
+
+                    if (args.CurrentValue != null)
+                    {
+                        Assert.IsType<int>(args.CurrentValue);
+                        setAutomatically[(int)args.CurrentValue]++;
+                    }
+                    else
+                    {
+                        Assert.IsType<int>(args.PreviousValue);
+                        unsetAutomatically[(int)args.PreviousValue]++;
+                    }
+
+                }
+            });
+
+            for (int i = 0; i < locals.Length; i++)
+            {
+                Assert.Equal(1, setManually[i]);
+                Assert.Equal(1, unsetAutomatically[i]);
+                Assert.Equal(0, setAutomatically[i]);
+            }
+
+            Array.Clear(setManually, 0, count);
+            Array.Clear(unsetAutomatically, 0, count);
+            Array.Clear(setAutomatically, 0, count);
+
+            ExecutionContext Default = ExecutionContext.Capture();
+
+            TestExecutionContextRun();
+
+            ExecutionContext.SuppressFlow();
+            try
+            {
+                TestExecutionContextRun();
+            }
+            finally
+            {
+                ExecutionContext.RestoreFlow();
+            }
+
+            void TestExecutionContextRun()
+            {
+                for (int i = 0; i < locals.Length; i++)
+                {
+                    ExecutionContext.Run(contexts[i].CreateCopy(), o =>
+                    {
+                        for (int index = 0; index < locals.Length; index++)
+                        {
+                            if (index >= i)
+                            {
+                                Assert.Equal(index, (int) locals[index].Value);
+                            }
+                            else
+                            {
+                                Assert.Null(locals[index].Value);
+                            }
+                        }
+
+                        ExecutionContext.Run(Default.CreateCopy(), _ =>
+                        {
+                            for (int index = 0; index < locals.Length; index++)
+                            {
+                                Assert.Null(locals[index].Value);
+                            }
+                        }, null);
+
+                        for (int l = 0; l < locals.Length; l++)
+                        {
+                            Assert.Equal(0, setManually[l]);
+                            Assert.Equal(l < i ? 0 : 1, unsetAutomatically[l]);
+                            Assert.Equal(l < i ? 0 : 2, setAutomatically[l]);
+                        }
+
+                        for (int c = 0; c < locals.Length; c++)
+                        {
+                            ExecutionContext.Run(contexts[c].CreateCopy(), _ =>
```

Comment:
It's possible I missed it, but do you explicitly test setting values within ExecutionContext.Run and making sure they've not propagated out?  That could/should be a separate test as well.

Reply:
Added